src/main/java/io/penguinstats/controller/v2/ReportControllerV2.java

```diff
@@ -0,0 +1,215 @@
+			for (int i = 0; i < dropsArray.length(); i++) {
+				JSONObject dropObj = dropsArray.getJSONObject(i);
+				Drop drop = new Drop(dropObj.getString("itemId"), dropObj.getInt("quantity"));
+				drops.add(drop);
+			}
+			if (furnitureNum > 0)
+				drops.add(new Drop("furni", furnitureNum));
+
+			// TODO: we should design checkers here, such as tag checker, time checker, limitation checker, etc.
+			Boolean isReliable = null;
+			if (source != null && source.equals("penguin-stats.io(internal)"))
+				isReliable = true;
+			else
+				isReliable = limitationUtil.checkDrops(drops, stageId, timestamp);
+			if (!isReliable)
+				logger.warn("Abnormal drop data!");
+
+			Integer times = 1;
+			// for gacha type stage, the # of times should be the sum of quantities.
+			Stage stage = stageService.getStageByStageId(stageId);
+			if (stage != null) {
+				Boolean isGacha = stage.getIsGacha();
+				if (isGacha != null && isGacha) {
+					times = 0;
+					for (Drop drop : drops) {
+						times += drop.getQuantity();
+					}
+				}
+			}
+
+			ItemDrop itemDrop = new ItemDrop(stageId, times, drops, timestamp, ip, isReliable, source, version, userID);
+			itemDropService.saveItemDrop(itemDrop);
+			String itemDropHashId = HashUtil.getHash(itemDrop.getId().toString());
+			return new ResponseEntity<>(itemDropHashId, HttpStatus.CREATED);
+		} catch (JSONException jsonException) {
+			logger.error("Error in saveSingleReport", jsonException);
+			return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+		} catch (Exception e) {
+			logger.error("Error in saveSingleReport", e);
+			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+		}
+	}
+
+	@ApiOperation("Get personal report history")
+	@GetMapping(path = "/history", produces = "application/json;charset=UTF-8")
+	public ResponseEntity<List<ItemDrop>> getPersonalReportHistory(HttpServletRequest request,
+			@RequestParam(name = "page", defaultValue = "0") Integer page,
+			@RequestParam(name = "page_size", defaultValue = "50") Integer pageSize,
+			@RequestParam(name = "sort_by", defaultValue = "timestamp") String sortBy,
+			@RequestParam(name = "direction", defaultValue = "ASC") String direction) {
+		try {
+			String userID = cookieUtil.readUserIDFromCookie(request);
+			if (userID == null) {
+				logger.error("Error in getPersonalReportHistory: Cannot read user ID");
+				return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+			}
+			logger.info("user " + userID + " GET /report/history\n");
+			Pageable pageable = PageRequest.of(page, pageSize, new Sort(Sort.Direction.fromString(direction), sortBy));
+			Page<ItemDrop> userItemDrops = itemDropService.getVisibleItemDropsByUserID(userID, pageable);
+			return new ResponseEntity<>(userItemDrops.getContent(), HttpStatus.OK);
+		} catch (Exception e) {
+			logger.error("Error in getPersonalReportHistory", e);
+			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+		}
+	}
+
+	@PostMapping(path = "/history")
```

Comment:
Suggest use HTTP Method `DELETE` instead of `POST`.

Reply:
fixed